### ISSUE TITLE
Prevent Type Widening on Dictionary Methods, like `dict.keys()`

### ISSUE THREAD

**Issue by jomido** [microsoft/pyright#338]
**Describe the bug**
When pulling, say, keys from a typed dictionary (`d.keys()`), the type applied to the key is lost (or rather, it's widened).

**To Reproduce**
Load the following code into/with a pyright-enabled context/tool:

```python
if __name__ == "__main__":

    Key = Literal['a', 'b', 'c']

    data: Dict[Key, str] = {
        'a': 'whatever',
        'b': 'whatever',
        'c': 'whatever',
    }

    keys: List[Key] = list(data.keys())  # type error

    key: Key = keys[0]  # type error
```

**Expected behavior**
There should be no type errors. 

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1469593/68998301-a2e46500-087e-11ea-9cfb-2bb6536d6dc8.png)

![image](https://user-images.githubusercontent.com/1469593/68998307-af68bd80-087e-11ea-8bb8-a8eb01b526d1.png)

![image](https://user-images.githubusercontent.com/1469593/68998310-b7286200-087e-11ea-83aa-6200285307f6.png)

**VS Code extension or command-line**
VS Code with `pyright 1.0.84`.

**Additional context**
This also happens with `d.values()`, `d.items()`, etc.

```python
if __name__ == "__main__":

    Key = Literal['a', 'b', 'c']
    Whatever = Literal['whatever']

    data: Dict[Key, Whatever] = {
        'a': 'whatever',
        'b': 'whatever',
        'c': 'whatever',
    }

    keys: List[Key] = list(data.keys())  #error

    key: Key = keys[0]  #error

    values: List[Whatever] = list(data.values())  #error

    value: Whatever = values[0]  #error

    keys_vals: List[Tuple[Key, Whatever]] = list(data.items())  #error

    key_val: Tuple[Key, Whatever] = key_vals[0]  #error
```

**Comment by erictraut** [comment]
Thanks for the bug report.

Pyright currently always strips literals when it performs TypeVar matching for generics. Failing to do so creates a bunch of problems in general. In cases where there is an expected type (because of an explicit type annotation), I might be able to retain the literals. I'll need to think about it more.

**Comment by erictraut** [comment]
Thanks again for reporting this. I think I've figured out a good solution, and the change will be in the next published version of pyright.

**Comment by jomido** [comment]
Amazing - thank you. I wasn’t sure if this was going to be possible. Looking forward to understanding both the problem (“[failing to strip Literals, when TypeVar matching for generics, creates a bunch of problems in general]”) and the fix.

**Comment by erictraut** [comment]
This is now implemented in 1.1.3, which I just published.
